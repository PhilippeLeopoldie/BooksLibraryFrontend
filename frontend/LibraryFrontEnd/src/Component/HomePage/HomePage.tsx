import "./HomePage.css";
import { BooksByGenre } from "./BooksByGenre/BooksByGenre";
import { GenreFilter } from "./GenreFilter/GenreFilter";
import { NewBooks } from "./NewBooks/NewBooks";
import { PopularBooks } from "./PopularBooks/PopularBooks";
import { useContext, useEffect, useState } from "react";
import { FilteredGenreContext, ThemeContext} from "../../App/App";

export const HomePage = () => {
    const theme = useContext(ThemeContext);
    const genreContext = useContext(FilteredGenreContext);
    const [numberOfBooks, setNumberOfBooks] = useState<number>(3);
    const [displayedContent, setDisplayedContent] = useState<JSX.Element>(<></>);


    useEffect(() => {
        if (genreContext?.genreFilter === 'Popular') {
            setDisplayedContent(
                <PopularBooks maxNumBooks={`${numberOfBooks}`} />
            );
        }
        else if (genreContext?.genreFilter !== 'Recent') {
            setDisplayedContent(
                <section className={`booksByGenre_container booksByGenre_container--${theme}`}>
                    <BooksByGenre sessionStorageName={"genreFiltered"} />
                </section>
            );
        } else {
            setDisplayedContent(
                <NewBooks />
            );
        }
    }, [genreContext?.genreFilter, theme])

    return (
        <>
            <div className={`BookList--${theme}--flex`}>
                <GenreFilter />
                {displayedContent}
            </div>
        </>
    );
};
