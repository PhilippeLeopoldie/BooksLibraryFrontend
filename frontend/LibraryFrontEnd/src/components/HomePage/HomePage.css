.BookList--black {
    background-color: inherit;
    
}

.BookList--black--flex {
    display: flex;
    flex-direction: column;
    background-color: inherit;
}

.BookList--black--flex > h1 {
        background-color: inherit;
        color: #ffffff;
}

.BookListTitle {
    text-align:left;
    margin-left:1rem;
    margin-top:1.5rem;
}

.Books__Loading--black {
    margin-top: 5rem;
    background-color: inherit;
    color: #ffffff;
}

.Books__Loading--natural {
    margin-top: 5rem;
    background-color: inherit;
    color: #000000;
}

.bookListContainer--flex {
    display: flex;
    overflow-x: auto;
    -webkit-overflow-scrolling: touch;
    background-color: transparent;
}


@media screen and (min-width: 700px) {
    .bookListContainer--flex, .BookList__mostPopular {
        grid-template-columns: repeat(2, auto);
        margin-bottom: 0%;
    }
}

@media screen and (min-width: 1024px) {
    .bookListContainer--flex {
        display: grid;
        justify-content: center;
    }

    .bookListContainer--flex {
        grid-template-columns: repeat(3, auto);
    }
}

@media screen and (min-width: 1300px) {
    .bookListContainer--flex {
        grid-template-columns: repeat(4, auto);
    }
}

@media screen and (min-width: 1588px) {
    .bookListContainer--flex {
        grid-template-columns: repeat(5, auto);
    }
}
