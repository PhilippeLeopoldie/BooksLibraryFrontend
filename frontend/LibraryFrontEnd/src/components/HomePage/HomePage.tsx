import "./HomePage.css";
import { BooksByGenre } from "./BooksByGenre/BooksByGenre";
import { GenreFilter } from "./GenreFilter/GenreFilter";
import { NewBooks } from "./NewBooks/NewBooks";
import { PopularBooks } from "./PopularBooks/PopularBooks";
import { useContext, useEffect, useState } from "react";
import { FilteredGenreContext, ThemeContext} from "../../App/App";

export const HomePage = () => {
    const theme = useContext(ThemeContext);
    const genreContext = useContext(FilteredGenreContext);
    const numberOfPopularBooks = 6;
    const [displayedContent, setDisplayedContent] = useState<JSX.Element>(<></>);


    useEffect(() => {
        if (genreContext?.genreFilter === 'Popular') {
            setDisplayedContent(
                <PopularBooks maxNumBooks={`${numberOfPopularBooks}`} />
            );
        }
        else if (genreContext?.genreFilter !== 'Recent') {
            setDisplayedContent(
                <section className={`booksByGenre_container booksByGenre_container--${theme}`}>
                    <BooksByGenre sessionStorageName={"genreFiltered"} />
                </section>
            );
        } else {
            setDisplayedContent(
                <NewBooks />
            );
        }
    }, [genreContext?.genreFilter, theme])

    return (
        <>
            <div className={`BookList--${theme}--flex`}>
                <GenreFilter />
                {displayedContent}
            </div>
        </>
    );
};
