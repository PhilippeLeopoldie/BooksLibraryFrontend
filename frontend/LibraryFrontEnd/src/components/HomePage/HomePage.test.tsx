import { HomePage } from "./HomePage";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { useState } from "react";


jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

type BookType = {
  book: {
    id: number;
    title: string;
    author: string;
  };
};
const book: BookType = {
  book: {
    id: 1,
    title: "title 1",
    author: "author 1",
  },
};
describe("BookList rendering", () => {
  test("snapShot: should render BookList", () => {
    (useState as jest.Mock).mockImplementation(() => [book, jest.fn()]);
    const {asFragment} = render(<HomePage/>)

    expect(asFragment()).toMatchSnapshot();
  })

  test("renders 'Loading...' when books state is null", () => {
    // arange
    (useState as jest.Mock).mockImplementation(() => [null, jest.fn()]);

    // Act
    render(<HomePage />);
    const loadingMessage = screen.getByText("Loading...");

    // expect
    expect(loadingMessage).toBeInTheDocument();
  });

  test("should not render 'Loading...' when books state is not null", () => {
    // arrange
   
    (useState as jest.Mock).mockImplementation(() => [book, jest.fn()]);

    // Act
    render(<HomePage />);
    const loadingMessage = screen.queryByText("Loading...");

    // expect
    expect(loadingMessage).toBeNull();
  });
});
