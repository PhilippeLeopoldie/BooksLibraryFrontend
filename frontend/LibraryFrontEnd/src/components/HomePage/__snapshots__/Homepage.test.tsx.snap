// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList rendering snapShot: should render BookList 1`] = `
<DocumentFragment>
  <header
    class="BookList--natural books__header--flex"
  >
    <h2
      class="BookList"
    >
      Recommendation of the day
    </h2>
    <h3
      class="BookList"
    />
  </header>
  <main
    class="bookListContainer"
  />
</DocumentFragment>
`;
